### PR TITLE
Code Review for Players Interaction and Computer Player Implementation

### PR DESCRIPTION
**Pull request by rjdeloss**
Based on what we spoke during IPM.

### CODE REVIEW

File: app/src/main/java/tttTDD/Java/Board.java [hunk 1]
```diff
@@ -1,44 +1,90 @@
...
-            this.board[i] = String.valueOf(i + 1);
-        }
+        this.board = initializeBoard(boardSize);
     }
 
-    public String[] getBoard() {
-        return this.board;
+    public String getSpace(int location) {
+        return isLocationValid(location) ? null : this.board[location];
     }
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:31]: Intentionally returning null values increases the risk of null pointer exceptions, because it's pretty common for consumers to forget that the method could always return (e.g.) int _or null_.

A fairly modern way to handle this is to use the [Optional single-value container](https://www.oracle.com/technical-resources/articles/java/java8-optional.html) introduced in Java 8. This would require the consumer to unwrap it to get to the value, forcing them to consider the negative case.

A simpler way would be to throw a (user-defined) exception. Consumers might ignore the exception annotation and naively use the returned value, but if so the null value will pass the exception up the stack until it halts the program, focusing attention on the buggy code, rather than creating hard to diagnose problems in methods that consume this.

File: app/src/main/java/tttTDD/Java/Board.java [hunk 2]
```diff
@@ -1,44 +1,90 @@
...
+        ArrayList<String> availableMovesArray = new ArrayList<>();
+
+        for (int i = 0; i < board.length; i++) {
+            if (isSpaceAvailable(i)) {
+                availableMovesArray.add(board[i]);
+            }
+        }
+
+        return Arrays.copyOf(availableMovesArray.toArray(), availableMovesArray.size(), String[].class);
+    }
```
Comment (nystromb) [inline app/src/main/java/tttTDD/Java/Board.java:88]: This is my opinion but I think it would be simpler/cleaner to use a stream here

```
return Arrays.stream(board).filter(this::isSpaceAvailable);
```

I believe this is functionally equivalent to what you're doing here

File: app/src/main/java/tttTDD/Java/Board.java [hunk 3]
```diff
@@ -1,44 +1,90 @@
...
-    public void updateBoard(int location, String marker) throws Exception {
+    public boolean updateBoard(int location, String marker) {
         if (isSpaceAvailable(location)) {
             this.board[location] = marker;
-        } else {
-            throw new Exception("The space is already occupied");
+            return true;
+        }
+
+        return false;
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:44]: As we discussed in IPM, the downside to merely returning `false` (instead of throwing) when given an illegal instruction is that it's pretty easy to consume this method in such a manner that you don't realize that it didn't do anything. The current implementation of `performMove` is a good example of this.

File: app/src/main/java/tttTDD/Java/Board.java [hunk 4]
```diff
@@ -1,44 +1,90 @@
...
+            }
+        }
+        return false;
+    }
+
+    private String[] initializeBoard(int boardSize) {
+        numberOfSpaces = boardSize * boardSize;
+        String[] board = new String[numberOfSpaces];
+        for (int i = 0; i < board.length; i++) {
+            board[i] = String.valueOf(i + 1);
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:60]: What benefits do we derive from having a different value in each uninhabited square?

File: app/src/main/java/tttTDD/Java/Human.java [hunk 5]
```diff
@@ -1,19 +1,18 @@
 package tttTDD.Java;
+import tttTDD.Java.Interfaces.Player;
 
-public class Player {
+public class Human implements Player {
     private String marker;
-    private boolean ai;
 
-    Player(String marker, boolean isComputer) {
+    Human(String marker) {
```
Comment (nystromb) [inline app/src/main/java/tttTDD/Java/Human.java:7]: 👍

File: app/src/main/java/tttTDD/Java/Board.java [hunk 6]
```diff
@@ -1,44 +1,90 @@
...
-import java.lang.reflect.Array;
+import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.List;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
 
 public class Board {
+    private final int DEFAULT_SIZE = 3;
     private final String[] board;
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:8]: On the same train of thought as [this comment](https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3/files#r874156496), are there any other data types that could more concisely communicate the state of the board?

If so, such a refactor would definitely make sense as it's own PR, just wanted to suggest the idea since this is the first time I'm reviewing code immediately adjacent to it.

File: app/src/main/java/tttTDD/Java/Game.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        currentPlayer = player1;
+        gameCompleted = false;
+    }
+    // humanPlayer.move(int) => return int
+    // list all the things that it does and all the things that it needs
+    public boolean performMove(int playerInput) {
+        if (!currentPlayer.isComputer()) {
+            board.updateBoard(playerInput, currentPlayer.getMarker());
+        } else {
+            board.updateBoard(board.getFirstAvailableMove(), currentPlayer.getMarker());
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Game.java:25]: The [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) is usually discussed in the context of the interfaces we build for end users, but I think it's good to apply quite generally.

So, for example, if I were consuming this method (and hadn't authored it), I would expect, based on the function signature and name, that the move I pass in is the one that is made, or to get a `false` return value, which I would assume implied that no move was made at all.

I would be quite astonished to find that actually a completely arbitrary move would instead be made every other time that I call this.

This can be partially alleviated by documenting the behavior (and proper use) of the method in Javadoc, but if I'm in a hurry, I'm likely to make an (incorrect) assumption that this is more straightforward than it is.

File: app/src/main/java/tttTDD/Java/Game.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+import tttTDD.Java.Interfaces.Player;
+
+public class Game {
+    Board board;
+    Player player1;
+    Player player2;
+    Player currentPlayer;
+    boolean gameCompleted;
+    String gameWinner;
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Game.java:11]: All of these fields are implicitly public to any class in the package. Is this intentional?

For example, is it desirable for methods outside of the `Game` to be able to change the value of `gameCompleted`?

File: app/src/main/java/tttTDD/Java/Interfaces/Player.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package tttTDD.Java.Interfaces;
+
+public interface Player {
+    String getMarker();
+    boolean isComputer();
+}
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Interfaces/Player.java:6]: The player interface seems fairly anemic.

- Are there any fields that any implementation should always have?
- Should `Players` be able to _do_ anything, other than identify their marker and android character?

File: app/src/test/java/tttTDD/Java/BoardTest.java [hunk 10]
```diff
@@ -8,23 +8,25 @@ public class BoardTest {
     Board board;
 
     @Test
-    public void getBoardInstanceWithCorrectLength() {
-        board = new Board();
+    public void getBoardInstanceWithCorrectDefaultLengthOf3() {
+        board = new Board(3);
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/BoardTest.java:12]: The name of this test implies that you're going to leave length unspecified, but then you do.

File: app/src/test/java/tttTDD/Java/BoardTest.java [hunk 11]
```diff
@@ -8,23 +8,25 @@ public class BoardTest {
...
 
-        Assert.assertEquals(board.getBoard().length, 9);
+        Assert.assertEquals(board.getSpace(0), "1");
+        Assert.assertEquals(board.getSpace(8), "9");
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/BoardTest.java:15]: How do these two assertions prove that the board is the correct size?

File: app/src/test/java/tttTDD/Java/BoardTest.java [hunk 12]
```diff
@@ -8,23 +8,25 @@ public class BoardTest {
...
         Assert.assertEquals(board.getSpace(15), "16");
     }
 
     @Test
     public void getBoardSpace() {
-        board = new Board();
+        board = new Board(3);
 
         Assert.assertEquals(board.getSpace(5), "6");
         Assert.assertNotEquals(board.getSpace(5), "5");
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/BoardTest.java:32]: This second assertion is a bit redundant, unless there's a chance that your getter method also changes the the value of the field.

File: app/src/test/java/tttTDD/Java/BoardTest.java [hunk 13]
```diff
@@ -33,17 +35,32 @@ public void getBoardSpace() {
...
+        Assert.assertNull(board.getSpace(16));
+    }
+
+    @Test
+    public void updateBoardSpaceWithNewValue(){
         board = new Board();
-        board.updateBoard(2, "X");
 
+        Assert.assertTrue(board.updateBoard(2, "X"));
         Assert.assertEquals(board.getSpace(2), "X");
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/BoardTest.java:50]: Is there any way to refactor the `Board` class such that the `Assert` section doesn't need to make any more calls to the board instance to prove this?

Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/BoardTest.java:50]: Is there any way to refactor the `Board` class such that the `Assert` section doesn't need to make any more calls to the board instance to prove this?

File: app/src/test/java/tttTDD/Java/BoardTest.java [hunk 14]
```diff
@@ -33,17 +35,32 @@ public void getBoardSpace() {
...
     }
 
-    @Test(expected = Exception.class)
-    public void updatingTheSameSpaceWillResultInAnErrorException() throws Exception {
+    @Test
+    public void updatingTheSameSpaceWillResultInAnErrorException() {
         board = new Board();
         board.updateBoard(3, "X");
         board.updateBoard(3, "O");
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/BoardTest.java:57]: This call is both unnecessary, and demonstrates the down side to getting rid of the exception. At first glance it appears as if you made a valid move here but in reality nothing happens.

File: app/src/test/java/tttTDD/Java/GameStateTest.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
+package tttTDD.Java;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class GameStateTest {
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/GameStateTest.java:6]: If I'm not mistaken, this no longer has any purpose

File: app/src/test/java/tttTDD/Java/GameTest.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+        game = new Game(gameConfig);
+
+        game.performMove(playerInput1);
+        game.performMove(2);
+        game.performMove(playerInput2);
+        game.performMove(7);
+        game.performMove(playerInput3);
+        game.performMove(9);
+
+        Assert.assertEquals(game.board.getSpace(4), game.player2.getMarker());
```
Comment (billwanjohi) [inline app/src/test/java/tttTDD/Java/GameTest.java:75]: The flow of this test is very hard to parse. I have to read the implementation of `performMove`, `getFirstAvailableMove`, and `getAvailableMoves` in order to understand _when_ the computer takes _what_ specific moves. And even at that point, I have to keep a mental tally of those moves line by line, peaking up at the definitions of `playerInputX` and remembering to ignore the values passed on every other call.

Comment (billwanjohi) [reply app/src/test/java/tttTDD/Java/GameTest.java:75]: A lot of this comes down to the design of the method under test, but even without changing that, you could make this clearer by adding comments to each step and/or ditching the playerInput variables.

File: app/src/main/java/tttTDD/Java/Board.java [hunk 17]
```diff
@@ -1,44 +1,90 @@
...
+    }
+
+    public boolean hasWinningSet() {
+        for (int[] set : winningSets) {
+            if (board[set[0]].equals(board[set[1]]) && board[set[1]].equals(board[set[2]])) {
+                return true;
+            }
+        }
+        return false;
+    }
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:54]: As we discussed in IPM, this method is currently unsafe to consume from any board size greater than 3, because it will confidently state the game was won or lost based on the marks in the first three rows and columns of the grid.

This should be fixed by either restricting the board size, throwing an error, returning something other than true or false, or implementing an algorithm that's scalable to all board sizes. I put these options roughly in order of the level of effort needed.

File: app/src/main/java/tttTDD/Java/Game.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    // humanPlayer.move(int) => return int
+    // list all the things that it does and all the things that it needs
+    public boolean performMove(int playerInput) {
+        if (!currentPlayer.isComputer()) {
+            board.updateBoard(playerInput, currentPlayer.getMarker());
+        } else {
+            board.updateBoard(board.getFirstAvailableMove(), currentPlayer.getMarker());
+        }
+        gameCompleted = gameHasWinner();
+        swap();
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Game.java:28]: Also discussed in IPM, the performMove cannot swap players if the last move did not take effect, or else you'll potentially have the same player take 2 consecutive moves.

**Review by billwanjohi — CHANGES_REQUESTED** [review]
I made a lot of comments, most of which are either suggestions for code quality improvement or hints at how you might think about evolving the application. I do _not_ expect you to implement all of them as part of this PR.

Distinct from that are a few logical issues that I think need to be addressed now in order to make this "safe" to merge, where safe means not burying bugs in easy to miss ways (kinda fuzzy). 

- `getBoardInstanceWithCorrectDefaultLengthOf3` needs assertion(s) that are sufficient to establish the board size (https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#discussion_r874199416)
- ~~hasWinningSet or the Board constructor need to be altered to avoid incorrect game-end assessments (https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#discussion_r874277896)~~
- ~~performMove cannot swap players when no move is performed (https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#discussion_r874282685)~~

File: app/src/main/java/tttTDD/Java/Board.java (hunk 2 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Board.java:88]: I tried using the snippet up top and tried modifying it so that isSpace receives the location parameter but it didn't play nice with it.
`Arrays.stream(board).filter(space -> isSpaceAvailable(Integer.parseInt(space) - 1))`

it was asking for the function to return `Stream<String>` when `String[]` is needed.

File: app/src/main/java/tttTDD/Java/Board.java (hunk 3 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Board.java:44]: When it comes to handling exceptions, would you say is best to have it happen at a higher level in the app or here within the `Board` class?

File: app/src/main/java/tttTDD/Java/Board.java (hunk 4 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Board.java:60]: The benefits derived is to distinguish each space from one another once a board is rendered if there's only a keyboard available to enter the appropriate cell.

File: app/src/main/java/tttTDD/Java/Board.java (hunk 6 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Board.java:8]: currently there's an object `GameConfiguration` that gets passed into the game to set the parameters. Eventually there will be an object that holds all the state variables within the `Game`.

File: app/src/main/java/tttTDD/Java/Interfaces/Player.java (hunk 9 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Interfaces/Player.java:6]: It should be this lean since it's shared between the `Computer` and `Human` player. Anything more would imply that it knows about the other classes. One of the questions I brought up during IPM whether the computer needed to know about the available moves and if it did, would that mean it also has access to the board.

File: app/src/test/java/tttTDD/Java/BoardTest.java (hunk 11 above)
Comment (rjdeloss) [reply app/src/test/java/tttTDD/Java/BoardTest.java:15]: It asserts that the lower and upper bounds of the board exist. I can always add another to specify it did not go over the  total number of cells for a 3x3 board.

File: app/src/test/java/tttTDD/Java/BoardTest.java (hunk 13 above)
Comment (rjdeloss) [reply app/src/test/java/tttTDD/Java/BoardTest.java:50]: The other way would be to compare the board data against an expected board state but we got rid of the method that calls `getBoard` because it had access to too much information per a conversation with @joeyshipley.

File: app/src/test/java/tttTDD/Java/GameTest.java (hunk 16 above)
Comment (rjdeloss) [reply app/src/test/java/tttTDD/Java/GameTest.java:75]: I do agree with you that it can be difficult to follow along. This was the result of testing the game conclusion without a game loop or direct player input from an external source like a `Scanner`

File: app/src/main/java/tttTDD/Java/Board.java (hunk 17 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Board.java:54]: It's currently in the works. I decided to write a validation that generates a dynamic set of winning moves and then compares them against those locations on the board. We'll see how it pans out once the idea is more developed.

File: app/src/main/java/tttTDD/Java/Board.java (hunk 3 above)
Comment (billwanjohi) [reply app/src/main/java/tttTDD/Java/Board.java:44]: It may be appropriate to handle (catch) the exception that I'm suggesting you throw here in other methods that call `updateBoard` within the Board class, but it should _not_ be caught in the same method that throws it, because then it's just an obfuscated if statement.

Furthermore, some methods that call another function that can throw an exception have no need to handle them. If there's nothing coherent to do in response to the failure of (e.g.) the move to be made, then it may be appropriate to let the error continue to propagate up the stack.

File: app/src/main/java/tttTDD/Java/Board.java (hunk 4 above)
Comment (billwanjohi) [reply app/src/main/java/tttTDD/Java/Board.java:60]: > distinguish each space from one another

Gotcha, the stringified counting numbers indicate the universe of allowed inputs that you envision eventually presenting in a text-based user interface.

If that's the case (that we're optimizing the storage of this to match its display), though, why does the updateBoard method take a position as a zero-based integer?

In other words, why are we doing translation of the more efficient and non-specific representation of a tic-tac-toe square on the inputs to this class, but not on the outputs?

File: app/src/main/java/tttTDD/Java/Interfaces/Player.java (hunk 9 above)
Comment (billwanjohi) [reply app/src/main/java/tttTDD/Java/Interfaces/Player.java:6]: > [Does] the computer [need] to know about the available moves and if it [does], would that mean it also has access to the board?

The answers to these questions flow from an earlier one, which is: what is the Computer (and relatedly the Player) actually supposed to _do_? In your current design and implementation, the answer is that Players do nothing at all, other than expose their (immutable) state to other objects. So, they also don't need to know or access anything external.

If instead you envisioned the Player class performing some action, then yes, it would likely reference your other classes, like Board.

File: app/src/main/java/tttTDD/Java/Game.java (hunk 7 above)
Comment (joeyshipley) [reply app/src/main/java/tttTDD/Java/Game.java:25]: Great share. I think that message has largely been lost lately, even with all the talk around mitigating/removing side effects as a whole.

File: app/src/main/java/tttTDD/Java/Game.java (hunk 8 above)
Comment (joeyshipley) [reply app/src/main/java/tttTDD/Java/Game.java:11]: I like the call out here around Game, the games state, and a GameState class. -- I would give pause to a need that extracts private instance variables into another class without specific behavior on the new class that conflicts with the origin place of the variables.

Comment (nystromb) [reply app/src/main/java/tttTDD/Java/Game.java:11]: I agree, I dont think we need to introduce a `GameState` object at this point. I think these should be explicitly private and if this information is needed, it should be encapsulated in a method

File: app/src/test/java/tttTDD/Java/BoardTest.java (hunk 11 above)
Comment (joeyshipley) [reply app/src/test/java/tttTDD/Java/BoardTest.java:15]: I think to fully imply the board size it would need to also show that out of bounds is invalid. In this case: -1 and 9 either return invalid board signifiers or throw an exception (depending on the practice of the codebase).

File: app/src/test/java/tttTDD/Java/BoardTest.java (hunk 14 above)
Comment (joeyshipley) [reply app/src/test/java/tttTDD/Java/BoardTest.java:57]: I don't disagree. But want to point out, that it is only flawed due to not consuming the response after switching the implementation to be response-based. I do not yet see it as an inherent flaw not using exceptions.

File: app/src/main/java/tttTDD/Java/Game.java (hunk 8 above)
Comment (joeyshipley) [reply app/src/main/java/tttTDD/Java/Game.java:11]: Ah, cool, didn't know Java did that. -- Let's specifically mark those as private then.

File: app/src/main/java/tttTDD/Java/Interfaces/Player.java (hunk 9 above)
Comment (billwanjohi) [reply app/src/main/java/tttTDD/Java/Interfaces/Player.java:6]: Yes. I'm sure there are other ways of pulling it off, but the most straightforward way of one object acting on another is for the first to call the second's public methods. That is distinct from the question of whether `Player` should be able to directly mutate the `Board`'s fields, which OOP discourages.

File: app/src/main/java/tttTDD/Java/Board.java [hunk 19]
```diff
@@ -85,6 +83,54 @@ private String[] getAvailableMoves(String[] board) {
...
+
+        for(int i = boardSize * 2 + 1; i < boardSize * 2 + 2; i++) {
+            for (int j = 0; j < boardSize; j++) {
+                winningSets[i][j] = winningSets[j][d--];
+
+            }
+        }
+
+        return winningSets;
+    }
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:135]: This method is very complex, enough so that I cannot understand what it is doing. Some things that would help tremendously:

- consider breaking up into several private methods with smaller purposes
- where breaking up a large function like this is not expedient, commenting the beginning of a new section of the subroutine can aid in comprehension
- use descriptive variable names. i and j at least have the benefit of being very conventional, but I have no idea what k and d are supposed to represent

File: app/src/main/java/tttTDD/Java/Board.java [hunk 20]
```diff
@@ -85,6 +83,54 @@ private String[] getAvailableMoves(String[] board) {
...
+            for(int j = 0; j < boardSize; j++) {
+                winningSets[i][j] = winningSets[j][d++];
+            }
+            d--;
+        }
+
+
+        for(int i = boardSize * 2 + 1; i < boardSize * 2 + 2; i++) {
+            for (int j = 0; j < boardSize; j++) {
+                winningSets[i][j] = winningSets[j][d--];
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:129]: > `[d--]`

Are you…incrementing this variable while simultaneously using it to access an array position?

File: app/src/main/java/tttTDD/Java/Board.java [hunk 21]
```diff
@@ -85,6 +83,54 @@ private String[] getAvailableMoves(String[] board) {
...
+
+        for(int i = boardSize * 2 ; i < boardSize * 2 + 1; i++) {
+            for(int j = 0; j < boardSize; j++) {
+                winningSets[i][j] = winningSets[j][d++];
+            }
+            d--;
+        }
+
+
+        for(int i = boardSize * 2 + 1; i < boardSize * 2 + 2; i++) {
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:127]: what do `boardSize * 2 + 1` and `boardSize * 2 + 2` represent? Consider assigning these as variables as an opportunity to document their meaning.

File: app/src/main/java/tttTDD/Java/Board.java [hunk 22]
```diff
@@ -85,6 +83,54 @@ private String[] getAvailableMoves(String[] board) {
...
+
+        for (int i = 0; i < set.length; i++) {
+            setBeingChecked[i] = board[set[i]];
+        }
+
+        return setBeingChecked;
+    }
+
+    private int[][] createWinningSet(int boardSize) {
+        int[][] winningSets = new int[(boardSize * 2) + 2][boardSize];
```
Comment (billwanjohi) [inline app/src/main/java/tttTDD/Java/Board.java:100]: Consider starting with a list of zero winners that you append to as you go along, rather than a list of null values that you replace as you go along.

File: app/src/test/java/tttTDD/Java/GameTest.java (hunk 16 above)
Comment (joeyshipley) [reply app/src/test/java/tttTDD/Java/GameTest.java:75]: Yeah, I can see how that would help for sure. Not something to stress at the moment, but being able to seed a current state of a board to test an individual move could be a good strategy to learn when dealing with complex series of events needed to get to a specific point in an app to test it. -- Something we can follow back upon at a later point, plenty to digest now as it is.

File: app/src/main/java/tttTDD/Java/Board.java (hunk 17 above)
Comment (rjdeloss) [reply app/src/main/java/tttTDD/Java/Board.java]: This was submitted today with a dynamic board and winning set validation.

File: app/src/main/java/tttTDD/Java/Game.java (hunk 18 above)
Comment (billwanjohi) [reply app/src/main/java/tttTDD/Java/Game.java]: good job on this

File: app/src/main/java/tttTDD/Java/Board.java (hunk 2 above)
Comment (billwanjohi) [reply app/src/main/java/tttTDD/Java/Board.java:87]: per https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#commitcomment-73948265 this can be left for another time

**Review by billwanjohi — CHANGES_REQUESTED** [review]
Only 1 of the 3 must-change comments I made are remaining: https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#discussion_r874199416. You already presented an idea that would resolve my issue, which just needs to be implemented: https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#discussion_r874731163

There are other things in here that are _important_, but not _blockers_ (most crucially: marking `Game` fields private), so Brian, Joey and I agree that they should be deferred to follow-up PRs.

**Review by nystromb — APPROVED** [review]

**Review by billwanjohi — APPROVED** [review]
